File: lib/data/models/knowledge.dart
import 'package:equatable/equatable.dart';

/// 知识库类型。
enum KnowledgeBaseType {
  /// 仿真校园演示资料
  demo('仿真校园知识库'),

  /// 用户导入资料
  user('用户导入知识库'),

  /// 演示 + 用户混合
  hybrid('混合知识库'),

  /// 空
  empty('空知识库');

  const KnowledgeBaseType(this.displayName);
  final String displayName;

  static KnowledgeBaseType fromString(String? value) {
    switch (value) {
      case 'demo':
        return KnowledgeBaseType.demo;
      case 'user':
        return KnowledgeBaseType.user;
      case 'hybrid':
        return KnowledgeBaseType.hybrid;
      default:
        return KnowledgeBaseType.empty;
    }
  }
}

/// 问答模式。
enum QaMode {
  /// 检索摘要(无 LLM 时使用)
  retrievalSummary('retrieval_summary', '检索摘要'),

  /// LLM RAG 生成
  llmRag('llm_rag', 'LLM RAG'),

  /// 无知识库依据
  noKnowledge('no_knowledge', '无知识库依据');

  const QaMode(this.code, this.displayName);
  final String code;
  final String displayName;

  static QaMode fromString(String? value) {
    switch (value) {
      case 'retrieval_summary':
        return QaMode.retrievalSummary;
      case 'llm_rag':
        return QaMode.llmRag;
      default:
        return QaMode.noKnowledge;
    }
  }
}

/// 索引状态。
enum IndexStatus { ready, empty, error }

/// 知识库状态信息 — 对齐后端 `KnowledgeStatus` schema。
class KnowledgeStatusInfo extends Equatable {
  const KnowledgeStatusInfo({
    required this.documentCount,
    required this.chunkCount,
    required this.indexStatus,
    required this.retrievalMethod,
    required this.isAvailable,
    required this.knowledgeBasePath,
    required this.knowledgeBaseType,
    required this.demoDocumentCount,
    required this.userDocumentCount,
    required this.llmAvailable,
    required this.qaMode,
    this.lastUpdated,
  });

  final int documentCount;
  final int chunkCount;
  final IndexStatus indexStatus;
  final String retrievalMethod;
  final bool isAvailable;
  final String knowledgeBasePath;
  final KnowledgeBaseType knowledgeBaseType;
  final int demoDocumentCount;
  final int userDocumentCount;
  final bool llmAvailable;
  final QaMode qaMode;
  final DateTime? lastUpdated;

  /// 是否包含演示资料(用于显示"仿真校园演示资料"声明)。
  bool get hasDemoDocuments => demoDocumentCount > 0;

  @override
  List<Object?> get props => [
        documentCount,
        chunkCount,
        indexStatus,
        retrievalMethod,
        isAvailable,
        knowledgeBasePath,
        knowledgeBaseType,
        demoDocumentCount,
        userDocumentCount,
        llmAvailable,
        qaMode,
        lastUpdated,
      ];
}

/// 文档摘要 — 对齐后端 `DocumentSummary` schema。
class KnowledgeDocumentSummary extends Equatable {
  const KnowledgeDocumentSummary({
    required this.documentId,
    required this.title,
    required this.contentHash,
    required this.isOfficial,
    required this.isExpired,
    required this.isDemo,
    required this.importedAt,
    this.sourceDepartment,
    this.sourceType,
    this.originalFilename,
    this.publishedAt,
    this.updatedAt,
    this.effectiveFrom,
    this.effectiveTo,
    this.version,
    this.applicableStudents,
    this.fileSize,
    this.fileExt,
  });

  final String documentId;
  final String title;
  final String contentHash;
  final bool isOfficial;
  final bool isExpired;
  final bool isDemo;
  final DateTime importedAt;
  final String? sourceDepartment;
  final String? sourceType;
  final String? originalFilename;
  final DateTime? publishedAt;
  final DateTime? updatedAt;
  final DateTime? effectiveFrom;
  final DateTime? effectiveTo;
  final String? version;
  final String? applicableStudents;
  final int? fileSize;
  final String? fileExt;

  /// 文档来源类型展示文本。
  String get sourceTypeLabel {
    switch (sourceType?.toLowerCase()) {
      case 'official':
        return '官方文件';
      case 'notice':
        return '校园通知';
      case 'guide':
        return '办事指南';
      case 'policy':
        return '制度文件';
      default:
        return sourceType ?? '未分类';
    }
  }

  /// 文档大小展示文本。
  String get fileSizeLabel {
    if (fileSize == null) return '';
    final kb = fileSize! / 1024;
    if (kb < 1024) return '${kb.toStringAsFixed(1)} KB';
    return '${(kb / 1024).toStringAsFixed(2)} MB';
  }

  @override
  List<Object?> get props => [
        documentId,
        title,
        contentHash,
        isOfficial,
        isExpired,
        isDemo,
        importedAt,
        sourceDepartment,
        sourceType,
        originalFilename,
        publishedAt,
        updatedAt,
        effectiveFrom,
        effectiveTo,
        version,
        applicableStudents,
        fileSize,
        fileExt,
      ];
}

/// 上传状态(用于 UI 状态机)。
enum UploadStatus {
  /// 等待上传
  idle,

  /// 上传中
  uploading,

  /// 正在解析
  parsing,

  /// 正在建立索引
  indexing,

  /// 上传完成
  completed,

  /// 上传失败
  failed,
}

/// 上传进度信息。
class UploadProgress extends Equatable {
  const UploadProgress({
    required this.status,
    this.message,
    this.errorCode,
    this.documentId,
  });

  final UploadStatus status;
  final String? message;
  final String? errorCode;
  final String? documentId;

  bool get isTerminal =>
      status == UploadStatus.completed || status == UploadStatus.failed;
  bool get isInProgress =>
      status == UploadStatus.uploading ||
      status == UploadStatus.parsing ||
      status == UploadStatus.indexing;

  @override
  List<Object?> get props => [status, message, errorCode, documentId];
}

/// 重建索引进度。
class RebuildProgress extends Equatable {
  const RebuildProgress({
    required this.inProgress,
    this.chunkCount,
    this.documentCount,
    this.error,
  });

  final bool inProgress;
  final int? chunkCount;
  final int? documentCount;
  final String? error;

  @override
  List<Object?> get props => [inProgress, chunkCount, documentCount, error];
}

/// 数据清理动作。
enum DataManagementAction {
  clearChatHistory('clear_chat_history', '清除聊天记录'),
  clearUserTasks('clear_user_tasks', '清除用户待办'),
  deleteUserDocuments('delete_user_documents', '删除用户导入的知识库文档'),
  restoreDemoDocuments('restore_demo_documents', '恢复仿真演示资料'),
  resetMockDemoData('reset_mock_demo_data', '重置 Mock 演示数据');

  const DataManagementAction(this.code, this.displayName);
  final String code;
  final String displayName;
}

/// 数据清理结果。
class DataManagementResult extends Equatable {
  const DataManagementResult({
    required this.action,
    required this.success,
    required this.message,
    this.affectedCount = 0,
  });

  final DataManagementAction action;
  final bool success;
  final String message;
  final int affectedCount;

  @override
  List<Object?> get props => [action, success, message, affectedCount];
}

/// 上传文档时的元数据(对应后端 Form 字段)。
///
/// 注意:此类是数据模型,放在 models 目录;[KnowledgeManagementService] 接口
/// 在 `service_interfaces.dart` 中通过 `import '../models/models.dart'` 引用。
class KnowledgeDocumentMetadata {
  const KnowledgeDocumentMetadata({
    this.title,
    this.sourceDepartment,
    this.sourceType,
    this.publishedAt,
    this.updatedAt,
    this.effectiveFrom,
    this.effectiveTo,
    this.version,
    this.applicableStudents,
    this.isOfficial = false,
  });

  final String? title;
  final String? sourceDepartment;
  final String? sourceType;
  final DateTime? publishedAt;
  final DateTime? updatedAt;
  final DateTime? effectiveFrom;
  final DateTime? effectiveTo;
  final String? version;
  final String? applicableStudents;
  final bool isOfficial;

  /// 转换为后端 Form 字段(String 化的日期)。
  Map<String, String?> toFormFields() {
    String? fmt(DateTime? d) => d?.toUtc().toIso8601String();
    return {
      'title': title,
      'source_department': sourceDepartment,
      'source_type': sourceType,
      'published_at': fmt(publishedAt),
      'updated_at': fmt(updatedAt),
      'effective_from': fmt(effectiveFrom),
      'effective_to': fmt(effectiveTo),
      'version': version,
      'applicable_students': applicableStudents,
      'is_official': isOfficial ? 'true' : 'false',
    };
  }
}

File: backend/app/api/routes/knowledge.py
"""知识库管理路由。"""
from __future__ import annotations

import os
import tempfile
from pathlib import Path
from typing import List, Optional

from fastapi import APIRouter, File, Form, UploadFile
from fastapi.responses import JSONResponse

from ...core.exceptions import AppException, DocumentNotFound, FileNameUnsafe
from ...core.logging import logger
from ...core.security import sanitize_filename
from ...schemas.knowledge import (
    DataManagementResponse,
    DeleteResponse,
    DocumentMetadata,
    DocumentSummary,
    KnowledgeStatus,
    RebuildResponse,
)
from ...services.container import get_container

router = APIRouter()


def _determine_knowledge_base_type(
    demo_count: int, user_count: int
) -> str:
    """根据演示/用户文档数量判定知识库类型。"""
    if demo_count == 0 and user_count == 0:
        return "empty"
    if demo_count > 0 and user_count == 0:
        return "demo"
    if demo_count == 0 and user_count > 0:
        return "user"
    return "hybrid"


def _determine_qa_mode(
    *, llm_available: bool, is_available: bool
) -> str:
    """根据 LLM 与知识库可用性判定问答模式。"""
    if not is_available:
        return "no_knowledge"
    if llm_available:
        return "llm_rag"
    return "retrieval_summary"


@router.get("/knowledge/status", response_model=KnowledgeStatus)
async def knowledge_status() -> KnowledgeStatus:
    container = get_container()
    container.retrieval.rebuild()
    last_imported = container.document_repository.latest_imported_at()
    chunk_count = container.retrieval.chunk_count
    doc_count = container.document_repository.count_documents()
    demo_count = container.document_repository.count_demo_documents()
    user_count = container.document_repository.count_user_documents()
    is_available = chunk_count > 0
    kb_type = _determine_knowledge_base_type(demo_count, user_count)
    llm_available = (
        container.llm is not None and container.settings.llm_available
    )
    qa_mode = _determine_qa_mode(
        llm_available=llm_available, is_available=is_available
    )
    return KnowledgeStatus(
        document_count=doc_count,
        chunk_count=chunk_count,
        last_updated=last_imported,
        index_status="ready" if is_available else "empty",
        retrieval_method="bm25",
        is_available=is_available,
        knowledge_base_path=str(container.settings.knowledge_base_dir),
        knowledge_base_type=kb_type,
        demo_document_count=demo_count,
        user_document_count=user_count,
        llm_available=llm_available,
        qa_mode=qa_mode,
    )


@router.get("/knowledge/documents", response_model=List[DocumentSummary])
async def list_documents() -> List[DocumentSummary]:
    container = get_container()
    docs = container.document_repository.list_documents()
    return [
        DocumentSummary(
            document_id=d.document_id,
            title=d.title,
            source_department=d.source_department,
            source_type=d.source_type,
            original_filename=d.original_filename,
            content_hash=d.content_hash,
            published_at=_parse_dt(d.published_at),
            updated_at=_parse_dt(d.updated_at),
            effective_from=_parse_dt(d.effective_from),
            effective_to=_parse_dt(d.effective_to),
            version=d.version,
            applicable_students=d.applicable_students,
            is_official=d.is_official,
            is_expired=d.is_expired,
            is_demo=d.is_demo,
            file_size=d.file_size,
            file_ext=d.file_ext,
            imported_at=_parse_dt(d.imported_at),
        )
        for d in docs
    ]


def _parse_dt(s: Optional[str]):
    if not s:
        return None
    from datetime import datetime
    try:
        return datetime.fromisoformat(s.replace("Z", "+00:00"))
    except ValueError:
        return None


@router.post("/knowledge/documents", response_model=DocumentSummary)
async def upload_document(
    file: UploadFile = File(...),
    title: Optional[str] = Form(None),
    source_department: Optional[str] = Form(None),
    source_type: Optional[str] = Form(None),
    published_at: Optional[str] = Form(None),
    updated_at: Optional[str] = Form(None),
    effective_from: Optional[str] = Form(None),
    effective_to: Optional[str] = Form(None),
    version: Optional[str] = Form(None),
    applicable_students: Optional[str] = Form(None),
    is_official: bool = Form(False),
) -> DocumentSummary:
    """上传文档到知识库。

    使用 multipart/form-data 上传文件，元数据通过 Form 字段提供。
    上传的文档默认 is_demo=False(用户导入资料)。
    """
    container = get_container()
    # 校验文件名(防止路径穿越、特殊字符)
    original_filename = file.filename or "upload"
    try:
        safe_filename = sanitize_filename(original_filename)
    except ValueError as e:
        raise AppException(
            str(e),
            code="FILE_NAME_UNSAFE",
            http_status=400,
        )
    # sanitize_filename 已剥离路径,用安全名重新取扩展名
    ext = Path(safe_filename).suffix.lstrip(".").lower()
    if ext not in container.settings.allowed_extensions_list:
        raise AppException(
            f"仅允许 {','.join(container.settings.allowed_extensions_list)} 类型",
            code="FILE_TYPE_NOT_ALLOWED",
            http_status=415,
        )
    # 读取到临时文件
    suffix = f".{ext}"
    tmp = tempfile.NamedTemporaryFile(delete=False, suffix=suffix)
    try:
        total = 0
        max_bytes = container.settings.max_upload_mb * 1024 * 1024
        while True:
            buf = await file.read(64 * 1024)
            if not buf:
                break
            total += len(buf)
            if total > max_bytes:
                tmp.close()
                os.unlink(tmp.name)
                raise AppException(
                    f"文件超过 {container.settings.max_upload_mb} MB 上限",
                    code="FILE_TOO_LARGE",
                    http_status=413,
                )
            tmp.write(buf)
        tmp.close()
        row, created = container.knowledge_ingestion.import_file(
            Path(tmp.name),
            original_filename=original_filename,
            source_department=source_department,
            source_type=source_type,
            published_at=published_at,
            updated_at=updated_at,
            effective_from=effective_from,
            effective_to=effective_to,
            version=version,
            applicable_students=applicable_students,
            is_official=is_official,
            is_demo=False,  # 上传的文档默认是用户导入,非演示资料
        )
        if not created:
            raise AppException(
                "文档已存在(哈希重复)",
                code="DOCUMENT_ALREADY_EXISTS",
                http_status=409,
                details={"document_id": row.document_id, "title": row.title},
            )
        # 若上传时未提供 title，使用文件推断的
        final_title = title or row.title
        # 重建索引
        container.retrieval.rebuild()
        return DocumentSummary(
            document_id=row.document_id,
            title=final_title,
            source_department=row.source_department,
            source_type=row.source_type,
            original_filename=row.original_filename,
            content_hash=row.content_hash,
            published_at=_parse_dt(row.published_at),
            updated_at=_parse_dt(row.updated_at),
            effective_from=_parse_dt(row.effective_from),
            effective_to=_parse_dt(row.effective_to),
            version=row.version,
            applicable_students=row.applicable_students,
            is_official=row.is_official,
            is_expired=row.is_expired,
            is_demo=row.is_demo,
            file_size=row.file_size,
            file_ext=row.file_ext,
            imported_at=_parse_dt(row.imported_at),
        )
    finally:
        try:
            os.unlink(tmp.name)
        except OSError:
            pass


@router.delete("/knowledge/documents/{document_id}", response_model=DeleteResponse)
async def delete_document(document_id: str) -> DeleteResponse:
    container = get_container()
    deleted = container.knowledge_ingestion.delete_document(document_id)
    if not deleted:
        raise DocumentNotFound(f"文档 {document_id} 不存在")
    container.retrieval.rebuild()
    return DeleteResponse(success=True, document_id=document_id)


@router.post("/knowledge/rebuild", response_model=RebuildResponse)
async def rebuild_index() -> RebuildResponse:
    container = get_container()
    n = container.knowledge_ingestion.rebuild_index()
    return RebuildResponse(
        success=True,
        document_count=container.document_repository.count_documents(),
        chunk_count=n,
        message="索引已重建",
    )


@router.post("/knowledge/manage/{action}", response_model=DataManagementResponse)
async def data_management(action: str) -> DataManagementResponse:
    """数据管理操作。

    支持的 action:
    - delete_user_documents: 删除所有用户导入的知识库文档(保留测试环境资料)
    - delete_all_documents: 删除所有知识库文档(包括测试环境资料)

    备注:
    - 不再提供"恢复演示资料 / 一键重置演示数据"等生产接口;
      测试环境资料仅在 dev/test 启动时通过 AUTO_IMPORT_DEMO 导入,
      production 已被 config 校验拦截。
    """
    container = get_container()
    action = action.strip().lower()
    if action == "delete_user_documents":
        n = container.knowledge_ingestion.delete_all_user_documents()
        container.retrieval.rebuild()
        return DataManagementResponse(
            success=True,
            action=action,
            affected_count=n,
            message=(
                f"已删除 {n} 份用户导入文档,测试环境资料保留。"
                if n > 0
                else "没有用户导入文档可删除。"
            ),
        )
    if action == "delete_all_documents":
        n = container.knowledge_ingestion.delete_all_documents()
        container.retrieval.rebuild()
        return DataManagementResponse(
            success=True,
            action=action,
            affected_count=n,
            message=f"已删除全部 {n} 份知识库文档(包括测试环境资料)。",
        )
    raise AppException(
        f"不支持的数据管理操作: {action}",
        code="INVALID_ACTION",
        http_status=400,
    )
